velox/expression/tests/FuzzerRunner.h

```diff
@@ -106,16 +106,139 @@ class FuzzerRunner {
     return output;
   }
 
-  static const std::unordered_map<
+  static std::unordered_map<
       std::string,
       std::vector<facebook::velox::exec::FunctionSignaturePtr>>
-      kSpecialForms;
+      specialFormSignatures;
+
+  /// Add supported signatures for the Cast special form for basic types.
+  /// TODO: Add supported Cast signatures to CastTypedExpr and expose them to
+  /// fuzzer instead of hard-coding signatures here.
+  static void addSignaturesForCast() {
+    std::vector<std::string> integralTypes{
+        "tinyint", "smallint", "integer", "bigint"};
+    std::vector<std::string> floatingPointTypes{"real", "double"};
+
+    std::vector<facebook::velox::exec::FunctionSignaturePtr> signatures;
+
+    // integral --> floating-point.
+    for (const auto& fromType : integralTypes) {
+      for (const auto& toType : floatingPointTypes) {
+        signatures.push_back(facebook::velox::exec::FunctionSignatureBuilder()
+                                 .argumentType(fromType)
+                                 .returnType(toType)
+                                 .build());
+      }
+    }
+
+    // floating-point --> integral.
+    for (const auto& fromType : floatingPointTypes) {
+      for (const auto& toType : integralTypes) {
+        signatures.push_back(facebook::velox::exec::FunctionSignatureBuilder()
+                                 .argumentType(fromType)
+                                 .returnType(toType)
+                                 .build());
+      }
+    }
+
+    // integral --> varchar.
+    for (const auto& fromType : integralTypes) {
+      signatures.push_back(facebook::velox::exec::FunctionSignatureBuilder()
+                               .argumentType(fromType)
+                               .returnType("varchar")
+                               .build());
+    }
+
+    // varchar --> integral.
+    for (const auto& toType : integralTypes) {
+      signatures.push_back(facebook::velox::exec::FunctionSignatureBuilder()
+                               .argumentType("varchar")
+                               .returnType(toType)
+                               .build());
+    }
+
+    // floating-point --> varchar.
+    for (const auto& fromType : floatingPointTypes) {
+      signatures.push_back(facebook::velox::exec::FunctionSignatureBuilder()
+                               .argumentType(fromType)
+                               .returnType("varchar")
+                               .build());
+    }
+
+    // floating-point --> floating-point.
+    for (const auto& fromType : floatingPointTypes) {
+      for (const auto& toType : floatingPointTypes) {
+        signatures.push_back(facebook::velox::exec::FunctionSignatureBuilder()
```

Comment:
Do we end up generating a  REAL -> REAL and DOUBLE -> DOUBLE signatures? Is this intentional? I remember seeing somewhere that such casts are not allowed.

Reply:
Yes, I checked that both Presto and Velox do support REAL -> REAL and DOUBLE -> DOUBLE. In addition, other integral types, VARCHAR, TIMESTAMP, and DATE do support casting to themselves. What you remember might be that I mentioned we don't support JSON -> JSON in Velox. I just found that Presto support JSON -
> JSON too, so we also need to fix this.